### PR TITLE
Testability

### PR DESCRIPTION
**Pull request by palango**
Based on #515

- Make some functions free, so that they can be used in tests.
- Use those functions to test the removal of old requests and claims

Resolves https://github.com/beamer-bridge/beamer/issues/356

### CODE REVIEW

**Review by weilbith — COMMENTED** [review]
Mistake...

File: beamer/chain.py [hunk 1]
```diff
@@ -183,97 +184,85 @@ def _process_events(self) -> None:
...
-                    to_remove.append(request.id)
 
-        for request_id in to_remove:
-            self._context.requests.remove(request_id)
+def _transact(func: web3.contract.ContractFunction, **kwargs: Any) -> web3.types.HexBytes:
+    try:
+        tx_hash = func.transact(cast(Optional[TxParams], kwargs))
+    except (web3.exceptions.ContractLogicError, requests.exceptions.RequestException) as exc:
+        raise _TransactionFailed() from exc
+    return tx_hash
```
Comment (weilbith) [inline beamer/chain.py:204]: Nitpick:

```suggestion
   try:
        return func.transact(cast(Optional[TxParams], kwargs))
    except (web3.exceptions.ContractLogicError, requests.exceptions.RequestException) as exc:
        raise _TransactionFailed() from exc
```

File: beamer/chain.py [hunk 2]
```diff
@@ -183,97 +184,85 @@ def _process_events(self) -> None:
             if not any_state_changed:
                 break
 
-    # TODO: pull this out of the event processor, so it can be tested independently
-    def _process_requests(self) -> None:
-        assert self._synced, "Not synced yet"
```
Comment (weilbith) [inline beamer/chain.py:188]: This line is missing from the new function. But honestly I don't even get why we added it, if the caller ensures we are synced anyways. So was that the reason you removed it?

File: beamer/chain.py [hunk 3]
```diff
@@ -183,97 +184,85 @@ def _process_events(self) -> None:
...
 
-    def __str__(self) -> str:
-        return "transaction failed: %s" % self.cause()
+        if claim.transaction_pending:
+            continue
 
-    def cause(self) -> str:
-        return str(self.__cause__)
+        block = context.request_manager.web3.eth.get_block("latest")
+        if claim.termination <= block["timestamp"]:
```
Comment (weilbith) [inline beamer/chain.py:258]: It was `latest_timestamp >= claim.termination` before and now `claim.termination <= block["timestamp"]`. So I think to match this you need to remove the `=`.

**Review by weilbith — COMMENTED** [review]
TESTS 🥳 
Nice job 👍🏾 

Depending on the outcome of [this comment](https://github.com/beamer-bridge/beamer/pull/515#discussion_r843986815) of the "parent" PR, the tests need to change. Though I'm optimistic that this will not be the case. 😆

File: beamer/chain.py [hunk 4]
```diff
@@ -183,97 +184,85 @@ def _process_events(self) -> None:
...
 
-    # TODO: pull this out of the event processor, so it can be tested independently
-    def _process_requests(self) -> None:
-        assert self._synced, "Not synced yet"
-        self._log.info("Processing requests", num_requests=len(self._context.requests))
-
-        to_remove = []
-        for request in self._context.requests:
-            if self._stop:
-                break
```
[the remark this answers is not in the document]
Comment (palango) [reply beamer/chain.py]: Yeah, sharp eyes. The issue this solved for me is that when the loop for processing requests and claims, I cannot stop the agent.

so we could pass in the event to make this possible again. do you thing this is worth it?

File: beamer/chain.py (hunk 3 above)
Comment (palango) [reply beamer/chain.py:258]: 👁️

File: beamer/chain.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (palango) [reply beamer/chain.py]: >  I would expect it to just take longer until the loop is done.

Exactly what I mean. But with some caching and other improvements this time should go down a lot.

**Review by weilbith — APPROVED** [review]
I think it should work for the moment. 👍🏾

**Pull request merged**